sketch/src/sketch_rnn/model.ts

```diff
@@ -0,0 +1,363 @@
+      const numUnits = this.numUnits;
+      const h = tf.tensor2d(state[1], [1, numUnits]);
+      const NOUT = this.N_MIXTURE;
+
+      const sqrttemp = tf.scalar(Math.sqrt(temp));
+      const softtemp = tf.scalar(discreteTemp);
+
+      const z = tf.add(tf.matMul(h, this.outputKernel), this.outputBias)
+        .squeeze();
+      const pen = tf.softmax(z.slice([0], [3]).div(softtemp));
+      const pi = tf.softmax(z.slice([3 + NOUT * 0], [NOUT]).div(softtemp));
+      const mu1 = z.slice([3 + NOUT * 1], [NOUT]);
+      const mu2 = z.slice([3 + NOUT * 2], [NOUT]);
+      const sigma1 = tf.exp(z.slice([3 + NOUT * 3], [NOUT])).mul(sqrttemp);
+      const sigma2 = tf.exp(z.slice([3 + NOUT * 4], [NOUT])).mul(sqrttemp);
+      const corr = tf.tanh(z.slice([3 + NOUT * 5], [NOUT]));
+
+      return [pi, mu1, mu2, sigma1, sigma2, corr, pen];
+    });
+    const result = [];
+    for (let i = 0; i < out.length; i++) {
+      result.push(out[i].dataSync());
+      out[i].dispose();
+    }
+    return result;
+  }
+
+  /**
+   * returns the zero/initial state of the model
+   *
+   * @returns zero state of the lstm: [c, h], where c and h are zero vectors.
+   */
+  zero_state() {
+    const result: Float32Array[] = [
+      new Float32Array(this.numUnits),
+      new Float32Array(this.numUnits),
+    ];
+    return result;
+  }
+
+  /**
+   * returns a new copy of the rnn state
+   *
+   * @param rnnState original [c, h] states of the lstm
+   *
+   * @returns copy of state of the lstm: [c, h]
+   */
+  copy_state(rnnState: Float32Array[]) {
+    const c = new Float32Array(rnnState[0]);
+    const h = new Float32Array(rnnState[1]);
+    return [c, h];
+  }
+
+  /**
+   * returns the zero input state of the model
+   *
+   * @returns [0, 0, 1, 0, 0].
+   */
+  zero_input() {
+    return [0, 0, 1, 0, 0];
+  }
+
+  /**
+   * Samples the next point of the sketch given pdf parameters
+   *
+   * @param pdf result from get_pdf() call
+   *
+   * @returns [dx, dy, pen_down, pen_up, pen_end]
+   */
+  sample(pdf: number[][]) {
+    // pdf is [z_pi, z_mu1, z_mu2, z_sigma1, z_sigma2, z_corr, z_pen]
+    // returns [x, y, eos]
+    const idx = support.sample_softmax(pdf[0]);
+    const mu1 = pdf[1][idx];
+    const mu2 = pdf[2][idx];
+    const sigma1 = pdf[3][idx];
+    const sigma2 = pdf[4][idx];
+    const corr = pdf[5][idx];
+    const penIdx = support.sample_softmax(pdf[6]);
+    const penstate = [0, 0, 0];
+    penstate[penIdx] = 1;
+    const delta = support.birandn(mu1, mu2, sigma1, sigma2, corr);
+    return [
+      delta[0] * this.scaleFactor,
+      delta[1] * this.scaleFactor,
+      penstate[0],
+      penstate[1],
+      penstate[2]
+    ];
+  }
+
+  /**
+   * Simplifies line using RDP algorithm
+   *
+   * @param line list of points [[x0, y0], [x1, y1], ...]
+   * @param tolerance (Optional) default 2.0
+   *
+   * @returns simpified line [[x0', y0'], [x1', y1'], ...]
+   */
+  simplify_line(line: number[][], tolerance?: number) {
```

Comment:
I know it's annoying to change, but function and variable names should all be camel-case according to style guidelines.

Reply:
Okay I'll convert to camelCase (ew..) and also change the demo code calls